### ISSUE TITLE
Accessing `arguments.callee.caller` in sloppy callback to JsBuiltIn throws

### ISSUE THREAD

**Issue by jackhorton** [Microsoft/ChakraCore#5642]
```js
[1, 2, 3].forEach(function () { arguments.callee.caller; })
```

This throws with Array.prototype.forEach as a JsBuiltIn, but passes with the native version. Same thing for .filter and .flatMap. V8 and SpiderMonkey do not throw.

**Comment by jackhorton** [comment]
cc @yullin-ms

**Comment by MSLaguana** [comment]
It looks like this is a function of the `forEach` argument being strict or not; you can replicate this kind of behavior very simply via
```js
function sloppy(f) { return f();}
function strict(f) { "use strict"; return f();}
function test() { arguments.callee.caller; }
sloppy(test);
strict(test);
```

The strict mode function throws, while the sloppy mode function doesn't.

**Comment by jackhorton** [comment]
Correct -- since it applies to library functions now with JsBuiltIns, though, I thought adding that would bring more attention to the issue :)

**Comment by rhuanjl** [comment]
Is the fix just to remove "use strict" from all the jsbuiltins that take callbacks or would that cause other problems?

**Comment by jackhorton** [comment]
I believe there were other issues with the fact that all built-in functions assumed psuedo-strict mode (they don't assume javascript semantics but strict mode emulates spec-ese better than sloppy mode, I think?)

@sigatrev or @meg-gupta may know why theyre all strict.

**Comment by sigatrev** [comment]
I don't think that strict mode is **strict**ly required for the built-ins. The real question is, are sloppy functions called by strict functions supposed to be considered strict or sloppy? If they are supposed to be sloppy, we should fix our strict-mode behavior and leave the built-ins as strict. If they are supposed to be strict, we should not use strict mode in any JsBuiltIns that makes callbacks.

running @MSLaguana's sample in other engines, V8 calls them as sloppy, and spidermonkey call them as strict like Chakra does.

@zenparsing can hopefully clear that up.

**Comment by pleath** [comment]
Strictness of the caller shouldn’t be a factor. Strict- or sloppiness is a static declarative property.

**Comment by jackhorton** [comment]
For what its worth, `xs` agrees with V8 on calling sloppy functions from a strict function. I can't get JSC to run, otherwise it would be the tiebreaker. Paul's point would make it seem like no matter which behavior is more popular, sloppy functions should always behave sloppily, regardless of their caller -- which sounds reasonable.

**Comment by pleath** [comment]
Look at it this way. There are early errors that apply only in strict mode.

**Comment by zenparsing** [comment]
The only rules here are at:

https://tc39.github.io/ecma262/#sec-forbidden-extensions

> If an implementation extends any function object with an own property named "caller" the value of that property, as observed using [[Get]] or [[GetOwnProperty]], must not be a strict function object. If it is an accessor property, the function that is the value of the property's [[Get]] attribute must never return a strict function when called.

So there isn't a spec violation here.

Since the functions in JsBuiltIn are strict, Chakra is installing some kind of throwing "caller" on the callback function, whereas v8 installs `null`.

**Comment by jackhorton** [comment]
Oh, interesting... I just tried with deleting Object.prototype and setting `false.x = 2` in the callback instead of accessing arguments.callee.caller. Everyone agrees that the callback takes sloppy `delete` semantics and returns false/silently fails in those other sloppy cases. Something is special about arguments.callee.caller, where we are incorrectly propagating strict semantics to the callback only in that case?

**Comment by fatcerberus** [comment]
Interestingly, V8 does not throw for the example given in the OP but *does* throw for:
```js
function foo() { 'use strict'; }
[1, 2, 3].forEach(function () { foo.caller; })
```
